week4/assignment/auth-practice/src/hooks/useJoin.ts

```diff
@@ -0,0 +1,149 @@
+  const [id, setId] = useState('');
+  const [password, setPassword] = useState('');
+  const [nickname, setNickname] = useState('');
+  const [phoneNumber, setPhoneNumber] = useState('');
+
+  // id, password, nickname, phoneNumber 에러 여부를 관리하는 state
+  const [idError, setIdError] = useState(false);
+  const [passwordError, setPasswordError] = useState(false);
+  const [nicknameError, setNicknameError] = useState(false);
+  const [phoneNumberError, setPhoneNumberError] = useState(false);
+
+  // id input이 변경될 때 호출되는 함수
+  const handleChangeId = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setId(e.target.value);
+  };
+  // password input이 변경될 때 호출되는 함수
+  const handleChangePassword = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setPassword(e.target.value);
+  };
+  // nickname input이 변경될 때 호출되는 함수
+  const handleChangeNickname = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setNickname(e.target.value);
+  };
+  // phoneNumber input이 변경될 때 호출되는 함수
+  const handleChangePhoneNumber = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setPhoneNumber(formatPhoneNumber(e.target.value));
+  };
+
+  // 회원가입 버튼 클릭 시 호출되는 함수
+  // id, password, nickname, phoneNumber가 유효한지 검사하는 함수
+  const validateJoin = () => {
+    // id, password, nickname, phoneNumber 에러 여부 초기화
+    setIdError(false);
+    setPasswordError(false);
+    setNicknameError(false);
+    setPhoneNumberError(false);
+
+    // id가 입력되지 않았을 경우 alert 출력 후 focus
+    if (!id) {
+      alert('아이디를 입력해주세요.');
+      if (idRef.current) {
+        idRef.current.focus();
+      }
+      setIdError(true);
+      return false;
+    }
+
+    // password가 입력되지 않았을 경우 alert 출력 후 focus
+    if (!password) {
+      alert('비밀번호를 입력해주세요.');
+      if (passwordRef.current) {
+        passwordRef.current.focus();
+      }
+      setPasswordError(true);
+      return false;
+    }
+
+    // password 형식이 맞지 않을 경우 alert 출력 후 focus
+    if (!validatePassword(password)) {
+      alert('비밀번호 형식이 맞지 않습니다.');
+      if (passwordRef.current) {
+        passwordRef.current.focus();
+      }
+      setPasswordError(true);
+      return false;
+    }
+
+    // nickname이 입력되지 않았을 경우 alert 출력 후 focus
+    if (!nickname) {
+      alert('닉네임을 입력해주세요.');
+      if (nicknameRef.current) {
+        nicknameRef.current.focus();
+      }
+      setNicknameError(true);
+      return false;
+    }
+
+    // phoneNumber가 입력되지 않았을 경우 alert 출력 후 focus
+    if (!phoneNumber) {
+      alert('전화번호를 입력해주세요.');
+      if (phoneNumberRef.current) {
+        phoneNumberRef.current.focus();
+      }
+      setPhoneNumberError(true);
+      return false;
+    }
+
+    return true;
+  };
+
+  // 회원가입 버튼 클릭 시 호출되는 함수
+  const handleClickJoin = async () => {
+    // id, password, nickname, phoneNumber가 유효하지 않을 경우 return
+    if (!validateJoin()) return;
+    const response = await postJoin({ authenticationId: id, password, nickname, phone: phoneNumber });
+    const { code, message } = response && response.data;
```

Comment:
p5: 요것도 아실 것 같지만, nullish 연산자 관련 링크도 첨부할게요!
저는 보통 요런식으로 쓰는 편이라서! 그냥 취향차이라고 생각하시면 될 것 같아요!
```suggestion
    const { code, message } = response?.data ?? {};
```
https://ko.javascript.info/nullish-coalescing-operator

Reply:
연산자에 대해선 알고 있었는데, 이런식으로 사용하는건 처음보네요! 다음엔 이방식도 사용해보겠습니다!

Reply:
도영님의 방법은 코드를 간결하게 줄이는데 도움이될 것 같네요!